rfcs/0180-broken-package-removal.md

```diff
@@ -0,0 +1,99 @@
+---
+feature: broken-package-removal
+start-date: 2024-07-14
+author: Jörg "Mic92" Thalheim
+co-authors:
+shepherd-team:
+shepherd-leader:
+related-issues:
+---
+
+# Summary
+
+[summary]: #summary
+
+This RFCs defines under what conditions we remove broken and unmaintained leaf packages.
+The RFC does not have the ambition to define all reasons to remove a package but
+instead focuses on simple rules that can be automated.
+Because this document is just the guideline for automatic package removal,
+it also leave it open, that a package can be removed earlier for other reasons.
+
+# Motivation
+
+[motivation]: #motivation
+
+Broken and unmaintained packages still cost us valuable time.
+Broken packages still have to be evaluated to some extent
+and unmaintained packages might still have to rebuild when one of their dependencies changes.
+Removing unmaintained/unused code hopefully increases the overall package quality and helps us
+to save resources.
+Because it is easy to add packages, it should be also easy to remove them, ideally automated.
+
+# Detailed design
+
+[design]: #detailed-design
+
+All packages marked as broken/unmaintained will be removed after a full NixOS release.
+For example `hello-unmaintained` will be marked as unmaintained in 23.11 then once we
+have 24.05, we can remove it from master.
+
+## Broken Packages
+
+Packages that are marked as broken unconditionally platforms should are canidates for removal.
+If the package had dependencies, those dependencies are also marked as indirectly broken.
+If they are not functional without the broken package, they should be removed as well.
+
+## Unmaintained packages
+
+We add an unmaintained warning for all packages, with an empty maintainer field that do not have packages depending on it.
```

Comment:
Would be nice if there was a point where the package is still in `nixpkgs` but it already fails the evaluation (just like `broken` does). That way people might find it easier to step up and maintain the package.

Reply:
This can be implemented by the `meta.problems` infra.

https://github.com/NixOS/rfcs/pull/180#issuecomment-2228888320

Reply:
This feels like a different RFC. We already have broken and use it here. This RFC does not conflict with having other flags that also indicate why something is broken.